### PR TITLE
feat(geom): Add support for multi-geometries [OL5]

### PR DESCRIPTION
**Pull request by samuel-girard**
Hello,

This change brings support for multi-geometries.  
The change is to be applied only after https://github.com/quentin-ol/ngx-openlayers/pull/183 is merged.  

There is just one **breaking change**: the coordinates input on `CollectionCoordinatesComponent` is not anymore limited to type `[number, number][]`, but it supports all types of coordinates.  
That means the existing code must be updated for polygons:  
Old style:
```html
<aol-feature>
    <aol-geometry-polygon>
        <aol-collection-coordinates
            [coordinates]="[[5, 45],[5.05, 45.05],[5.05, 44.95],[4.95, 44.95]]"
            [srid]="'EPSG:4326'"
        >
        </aol-collection-coordinates>
    </aol-geometry-polygon>
    <aol-style>
        <aol-style-stroke [color]="'red'"></aol-style-stroke>
        <aol-style-fill [color]="[255,0,0,0.5]"></aol-style-fill>
    </aol-style>
</aol-feature>
```
New style:  
```html
<aol-feature>
    <aol-geometry-polygon>
        <aol-collection-coordinates
            [coordinates]="[[[5, 45],[5.05, 45.05],[5.05, 44.95],[4.95, 44.95]]]"
            [srid]="'EPSG:4326'"
        >
        </aol-collection-coordinates>
    </aol-geometry-polygon>
    <aol-style>
        <aol-style-stroke [color]="'red'"></aol-style-stroke>
        <aol-style-fill [color]="[255,0,0,0.5]"></aol-style-fill>
    </aol-style>
</aol-feature>
```
Notice the `[coordinates]` input is now a `[number, number][][]`, as defined in GeoJSON.  

This also allows to display polygon with holes, which is not possible with the current code.

### CODE REVIEW

**Comment by Neonox31** [comment]
@samuel-girard I fixed the linting problem on next branch, could you rebase please ?

**Comment by Yakoust** [comment]
Hi all, 

I rebase this branch on next after merging OL5 PR.
This introduce a breaking change.
It LGTM but I would like to have another review

@quentin-ol @kekel87 @dabbid

File: projects/ngx-openlayers/src/lib/coordinate.component.ts [hunk 1]
```diff
@@ -81,59 +96,82 @@ export class CoordinateComponent implements OnChanges {
     <div class="aol-collection-coordinates"></div>
   `,
 })
-export class CollectionCoordinatesComponent implements OnChanges {
+export class CollectionCoordinatesComponent implements OnChanges, OnInit {
```
Comment (kekel87) [inline projects/ngx-openlayers/src/lib/coordinate.component.ts]: I would separate these components into two files. Now that they are bigger.

File: projects/ngx-openlayers/src/lib/coordinate.component.ts [hunk 2]
```diff
@@ -46,18 +48,31 @@ export class CoordinateComponent implements OnChanges {
...
-    let referenceProjection: Projection;
-    let referenceProjectionCode: string;
-    let transformedCoordinates: number[];
+    this.transformCoordinates();
+  }
 
-    referenceProjection = this.map.instance.getView().getProjection();
-    referenceProjectionCode = referenceProjection ? referenceProjection.getCode() : 'EPSG:3857';
+  private onMapViewChanged(event) {
+    this.mapSrid = (</*View*/ any>(</*BaseObject*/ any>event.target).get(event.key)).getProjection().getCode();
```
Comment (kekel87) [inline projects/ngx-openlayers/src/lib/coordinate.component.ts]: What is this syntax?

File: projects/ngx-openlayers/src/lib/coordinate.component.ts [hunk 3]
```diff
@@ -81,59 +96,82 @@ export class CoordinateComponent implements OnChanges {
...
   constructor(
     private map: MapComponent,
     @Optional() geometryLinestring: GeometryLinestringComponent,
-    @Optional() geometryPolygon: GeometryPolygonComponent
+    @Optional() geometryPolygon: GeometryPolygonComponent,
+    @Optional() geometryMultipoint: GeometryMultiPointComponent,
+    @Optional() geometryMultilinestring: GeometryMultiLinestringComponent,
+    @Optional() geometryMultipolygon: GeometryMultiPolygonComponent
   ) {
     // console.log('creating aol-collection-coordinates');
```
Comment (kekel87) [inline projects/ngx-openlayers/src/lib/coordinate.component.ts]: Remove comment.

File: projects/ngx-openlayers/src/lib/coordinate.component.ts [hunk 4]
```diff
@@ -81,59 +96,82 @@ export class CoordinateComponent implements OnChanges {
...
   ngOnChanges(changes: SimpleChanges) {
-    let referenceProjection: Projection;
-    let referenceProjectionCode: string;
-    let transformedCoordinates: Array<Coordinate>;
+    this.transformCoordinates();
+  }
 
-    // console.log('coordinates change: ', this.coordinates);
+  private onMapViewChanged(event) {
+    this.mapSrid = (</*View*/ any>(</*BaseObject*/ any>event.target).get(event.key)).getProjection().getCode();
```
Comment (kekel87) [inline projects/ngx-openlayers/src/lib/coordinate.component.ts]: What is this syntax?

File: projects/ngx-openlayers/src/lib/geom/geometrylinestring.component.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    <ng-content></ng-content>
+  `,
+})
+export class GeometryLinestringComponent extends SimpleGeometryComponent implements OnInit, OnDestroy {
+  public componentType = 'geometry-linestring';
+  public instance: LineString;
+
+  constructor(map: MapComponent, host: FeatureComponent) {
+    super(map, host);
+    // console.log('instancing aol-geometry-linestring');
```
Comment (kekel87) [inline projects/ngx-openlayers/src/lib/geom/geometrylinestring.component.ts]: Remove this comment

File: projects/ngx-openlayers/src/lib/geom/geometrymultilinestring.component.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    <ng-content></ng-content>
+  `,
+})
+export class GeometryMultiLinestringComponent extends SimpleGeometryComponent implements OnInit, OnDestroy {
+  public componentType = 'geometry-multilinestring';
+  public instance: MultiLineString;
+
+  constructor(map: MapComponent, host: FeatureComponent) {
+    super(map, host);
+    // console.log('creating aol-geometry-multilinestring');
```
Comment (kekel87) [inline projects/ngx-openlayers/src/lib/geom/geometrymultilinestring.component.ts]: Remove this comment

File: projects/ngx-openlayers/src/lib/geom/geometrymultipoint.component.ts [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    <ng-content></ng-content>
+  `,
+})
+export class GeometryMultiPointComponent extends SimpleGeometryComponent implements OnInit, OnDestroy {
+  public componentType = 'geometry-multipoint';
+  public instance: MultiPoint;
+
+  constructor(map: MapComponent, host: FeatureComponent) {
+    super(map, host);
+    // console.log('creating aol-geometry-multipoint');
```
Comment (kekel87) [inline projects/ngx-openlayers/src/lib/geom/geometrymultipoint.component.ts]: Remove this comment

File: projects/ngx-openlayers/src/lib/geom/geometrymultipolygon.component.ts [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    <ng-content></ng-content>
+  `,
+})
+export class GeometryMultiPolygonComponent extends SimpleGeometryComponent implements OnInit, OnDestroy {
+  public componentType = 'geometry-multipolygon';
+  public instance: MultiPolygon;
+
+  constructor(map: MapComponent, host: FeatureComponent) {
+    super(map, host);
+    // console.log('creating aol-geometry-multipolygon');
```
Comment (kekel87) [inline projects/ngx-openlayers/src/lib/geom/geometrymultipolygon.component.ts]: Remove this comment

File: projects/ngx-openlayers/src/lib/geom/geometrypoint.component.ts [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+    <ng-content></ng-content>
+  `,
+})
+export class GeometryPointComponent extends SimpleGeometryComponent implements OnInit, OnDestroy {
+  public componentType = 'geometry-point';
+  public instance: Point;
+
+  constructor(map: MapComponent, host: FeatureComponent) {
+    super(map, host);
+    // console.log('creating aol-geometry-point');
```
Comment (kekel87) [inline projects/ngx-openlayers/src/lib/geom/geometrypoint.component.ts]: Remove this comment

File: projects/ngx-openlayers/src/lib/geom/geometrypolygon.component.ts [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    <ng-content></ng-content>
+  `,
+})
+export class GeometryPolygonComponent extends SimpleGeometryComponent implements OnInit, OnDestroy {
+  public componentType = 'geometry-polygon';
+  public instance: Polygon;
+
+  constructor(map: MapComponent, host: FeatureComponent) {
+    super(map, host);
+    // console.log('creating aol-geometry-polygon');
```
Comment (kekel87) [inline projects/ngx-openlayers/src/lib/geom/geometrypolygon.component.ts]: Remove this comment

File: projects/ngx-openlayers/src/lib/geom/simplegeometry.component.ts [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+
+  constructor(protected map: MapComponent, protected host: FeatureComponent) {}
+
+  ngOnInit() {
+    this.host.instance.setGeometry(this.instance);
+  }
+
+  ngOnDestroy() {
+    // this.host.instance.setGeometry(null);
+  }
```
Comment (kekel87) [inline projects/ngx-openlayers/src/lib/geom/simplegeometry.component.ts]: Remove empty `ngOnDestroy()` and `implements OnDestroy`

File: src/app/draw-polygon/draw-polygon.component.ts [hunk 12]
```diff
@@ -86,5 +86,6 @@ export class DrawPolygonComponent implements OnInit {
         coordinates: [olGeomPolygon.getCoordinates()[0]],
       },
     };
+    console.log(this.feature);
```
Comment (kekel87) [inline src/app/draw-polygon/draw-polygon.component.ts]: Remove console.log.

**Comment by Yakoust** [comment]
@kekel87 agree with changes?

**Comment by kekel87** [comment]
@Yakoust Yep i agree :+1:

**Pull request merged**